ipsec/key_sockets.c

```diff
@@ -0,0 +1,338 @@
+
+
+static handle_t key_sockets_mutex;
+static key_socket_t *key_sockets[NUM_KEY_SOCKETS];
+
+
+static key_socket_t **tryget_key_socket(int fd)
+{
+	int i = fd - LWIP_KEY_SOCKET_OFFSET;
+
+	if (i < 0 || i >= NUM_KEY_SOCKETS)
+		return NULL;
+
+	return &key_sockets[i];
+}
+
+
+static key_socket_t **get_key_socket(int fd)
+{
+	key_socket_t **sock_ptr = tryget_key_socket(fd);
+
+	if (sock_ptr == NULL) {
+		errno = EBADF;
+		return NULL;
+	}
+
+	return sock_ptr;
+}
+
+
+static int alloc_key_socket(key_socket_t *sock)
+{
+	unsigned int i;
+
+	mutexLock(key_sockets_mutex);
+	for (i = 0; i < NUM_KEY_SOCKETS; ++i) {
+		if (!key_sockets[i]) {
+			key_sockets[i] = sock;
+			mutexUnlock(key_sockets_mutex);
+			return i + LWIP_KEY_SOCKET_OFFSET;
+		}
+	}
+	mutexUnlock(key_sockets_mutex);
+
+	return -1;
+}
+
+
+static void free_key_socket(key_socket_t **sock_ptr)
+{
+	mutexLock(key_sockets_mutex);
+	*sock_ptr = NULL;
+	mutexUnlock(key_sockets_mutex);
+}
+
+
+static key_socket_t *create_key_socket(void)
+{
+	key_socket_t *sock = malloc(sizeof(key_socket_t));
+	if (sock == NULL)
+		return NULL;
+
+	sock->mbuff = mbuff_alloc();
+	if (sock->mbuff == NULL) {
+		free(sock);
+		return NULL;
+	}
+
+	if (mutexCreate(&sock->mutex) < 0) {
+		mbuff_free(sock->mbuff);
+		free(sock);
+		return NULL;
+	}
+
+	if (semaphoreCreate(&sock->semaphore, 0) < 0) {
+		resourceDestroy(sock->mutex);
+		mbuff_free(sock->mbuff);
+		free(sock);
+		return NULL;
+	}
+
+	return sock;
+}
+
+
+static void destroy_key_socket(key_socket_t *sock)
+{
+	if (sock == NULL)
+		return;
+
+	semaphoreDone(&sock->semaphore);
+	resourceDestroy(sock->mutex);
+	mbuff_free(sock->mbuff);
+	free(sock);
+}
+
+
+void key_sockets_init(void)
+{
+	memset(key_sockets, 0, sizeof(key_sockets));
+	mutexCreate(&key_sockets_mutex);
+}
+
+
+int is_key_sockets_fd(int sockfd)
+{
+	key_socket_t **sock_ptr = tryget_key_socket(sockfd);
+	return (sock_ptr != NULL);
+}
+
+
+int key_sockets_socket(int domain, int type, int protocol)
+{
+	int fd;
+	key_socket_t *sock;
+
+	if (domain != AF_KEY) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	if (type != SOCK_RAW) {
+		errno = EPROTOTYPE;
+		return -1;
+	}
+
+	sock = create_key_socket();
+	if (sock == NULL) {
+		errno = ENOMEM;
+		return -1;
+	}
+
+	fd = alloc_key_socket(sock);
+	if (fd < 0) {
+		destroy_key_socket(sock);
+		errno = ENOMEM;
+		return -1;
+	}
+
+	return fd;
+}
+
+
+int key_sockets_close(int fd)
+{
+	key_socket_t **sock_ptr = get_key_socket(fd);
+	if (sock_ptr == NULL)
+		return -1;
+
+	destroy_key_socket(*sock_ptr);
+	free_key_socket(sock_ptr);
+
+	return 0;
+}
+
+
+int key_sockets_send(int sockfd, const void *buf, size_t len, int flags)
+{
+	key_socket_t **sock_ptr;
+	void *buf_cpy;
+	void *reply;
+	int ret;
+	size_t reply_len;
+	unsigned int i;
+
+	sock_ptr = tryget_key_socket(sockfd);
+	if (sock_ptr == NULL || *sock_ptr == NULL) {
+		errno = EBADF;
+		return -1;
+	}
+
+	/* FIXME: copy needed because there is some memory access issue on buf - investigate */
```

Comment:
`buf` alignment perhaps? It is cast to a struct inside of ipsec_sadbDispatch

Reply:
Alignment seems to be fine, without copy it crashes instantly.

Reply:
Yeah, that really sounds like unaligned `buf` - memcpy does not care about alignment, so it works just fine.

Anyway, when casting void * to some struct it is more kosher to use memcpy instead of forced cast, so maybe this is fine as is